core/utils/toolbox.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Utility functions for the toolbox and flyout.
+ * @author aschmiedt@google.com (Abby Schmiedt)
+ */
+'use strict';
+
+goog.provide('Blockly.utils.toolbox');
+
+
+/**
+ * A block definition.
+ * @typedef {{
+ *            kind:string,
+ *            blockxml:(string|Node)
+ *          }}
+ */
+Blockly.utils.toolbox.BlockInfo;
+
+/**
+ * A separator definition.
+ * @typedef {{
+ *            kind:string,
+ *            gap:?number
+ *          }}
+ */
+Blockly.utils.toolbox.SeparatorInfo;
+
+/**
+ * A button definition.
+ * @typedef {{
+ *            kind:string,
+ *            text:string,
+ *            callbackkey:string
+ *          }}
+ */
+Blockly.utils.toolbox.ButtonInfo;
+
+/**
+ * A label definition.
+ * @typedef {{
+ *            kind:string,
+ *            text:string
+ *          }}
+ */
+Blockly.utils.toolbox.LabelInfo;
+
+/**
+ * A toolbox item.
+ * @typedef {Blockly.utils.toolbox.BlockInfo|
+ *           Blockly.utils.toolbox.SeparatorInfo|
+ *           Blockly.utils.toolbox.ButtonInfo|
+ *           Blockly.utils.toolbox.LabelInfo}
+ */
+Blockly.utils.toolbox.ToolboxInfo;
+
+/**
+ * A category definition.
+ * @typedef {{
+ *            kind:string,
+ *            name:string,
+ *            categorystyle:?string,
+ *            colour:?string,
+ *            contents:Array.<Blockly.utils.toolbox.ToolboxInfo>
+ *          }}
+ */
+Blockly.utils.toolbox.CategoryInfo;
+
+/**
+ * Parse the provided toolbox definition into a consistent format.
+ * @param {Array|Node|NodeList|null} toolboxDef The definition of the
+ *    toolbox in one of its many forms.
+ * @return {!Array.<Blockly.utils.toolbox.ToolboxInfo>} Array of JSON holding
+ *    information on toolbox contents.
+ * @package
+ */
+Blockly.utils.toolbox.parseToolboxContents = function(toolboxDef) {
+  if (!toolboxDef) {
+    return [];
+  }
+  // The array can be either an array of xml or an array of JSON.
+  if (Array.isArray(toolboxDef)) {
+    if (toolboxDef.length && !(toolboxDef[0].nodeType)) {
+      return /** @type {!Array.<Blockly.utils.toolbox.ToolboxInfo>} */ (toolboxDef);
+    }
+  }
+
+  return Blockly.utils.toolbox.toolboxXmlToJson_(toolboxDef);
+};
+
+/**
+ * Convert the xml for a toolbox to JSON.
+ * @param {!NodeList|!Node|!Array.<Blockly.utils.toolbox.ToolboxInfo>} toolboxDef The
```

Comment:
Based on the types in the previous function, this can't be an array of ToolboxInfo.

Reply:
done